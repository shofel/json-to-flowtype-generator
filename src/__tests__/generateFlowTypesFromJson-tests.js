const generateFlowTypesFromJson = require('../generateFromTypesFromJson');

describe('generateFlowTypesFromJson', () => {
  it('should print JSON to a valid Flow type', () => {
    expect(
      generateFlowTypesFromJson(
        {
          testString: 'test',
          testBool: false,
          testNum: 123,
          testNull: null,
          testObj: {
            str: 'hello',
            bool: true,
            num: 12324,
            not: null
          },
          testObjList: [
            {
              str: 'hello',
              bool: true,
              num: 12324,
              not: null
            }
          ],
          testPrimitiveList: [123, 'test']
        },
        {
          name: 'TestType'
        }
      )
    ).toMatchSnapshot();
  });

  it('should find lists of objects, remove duplicates, unify all objects to a single object, and add maybe types to keys not present in all objects', () => {
    expect(
      generateFlowTypesFromJson(
        {
          testList: [
            { test: '123' },
            { test: '456', maybeProp: 123 },
            {
              test: '789',
              anotherMaybeProp: 'testing',
              maybeObj: { test: '123' }
            },
            {
              test: '789',
              nestedList: [
                { prop: 'test', maybeProp: 123456 },
                { prop: 'test' }
              ]
            }
          ]
        },
        {
          name: 'TestType'
        }
      )
    ).toMatchSnapshot();
  });

  it('should handle nested arrays that are sometimes empty', () => {
    expect(
      generateFlowTypesFromJson(
        {
          testList: [
            { test: '123', nestedList: [] },
            { test: '456', maybeProp: 123, nestedList: [] },
            {
              test: '789',
              anotherMaybeProp: 'testing',
              maybeObj: { test: '123' },
              nestedList: []
            },
            {
              test: '789',
              nestedList: [
                { prop: 'test', maybeProp: 123456 },
                { prop: 'test' }
              ]
            }
          ]
        },
        {
          name: 'TestType'
        }
      )
    ).toMatchSnapshot();
  });

  describe('Realistic examples', () => {
    describe('API responses', () => {
      test('api response should yield a sensible Flow type', () => {
        expect(
          generateFlowTypesFromJson(
            {
              data: {
                notifications: {
                  count: 123,
                  edges: [
                    {
                      cursor: '1234',
                      node: {
                        id: 1234,
                        message: 'hejsan'
                      }
                    },
                    {
                      cursor: '1234',
                      node: {
                        id: 1234,
                        message: null
                      }
                    }
                  ]
                },
                user: {
                  id: 1234,
                  firstName: 'Mrs',
                  lastName: 'Test',
                  email: 'test@test.com',
                  friends: [
                    {
                      id: 1234,
                      firstName: 'Friend',
                      lastName: 'One',
                      email: 'friend@friend.com',
                      isOnline: true,
                      lastMessaged: 123455
                    },
                    {
                      id: 1234,
                      firstName: 'Friend',
                      lastName: 'One',
                      email: 'friend@friend.com',
                      isOnline: false,
                      lastMessaged: null
                    },
                    {
                      id: 1234,
                      firstName: 'Friend',
                      lastName: 'One'
                    }
                  ]
                }
              }
            },
            {
              name: 'UserAPIResponse'
            }
          )
        ).toMatchSnapshot();
      });
    });
  });
});
