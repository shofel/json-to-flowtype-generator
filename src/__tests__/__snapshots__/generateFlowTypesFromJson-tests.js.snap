// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateFlowTypesFromJson Realistic examples API responses api response should yield a sensible Flow type 1`] = `
"type UserAPIResponse = {|
  +data: {|
    +notifications: {|
      +count: number,
      +edges: $ReadOnlyArray<{|
        +cursor: string,
        +node: {|
          +id: number,
          +message: ?string
        |}
      |}>
    |},
    +user: {|
      +id: number,
      +firstName: string,
      +lastName: string,
      +email: string,
      +friends: $ReadOnlyArray<{|
        +id: number,
        +firstName: string,
        +lastName: string,
        +email: ?string,
        +isOnline: ?boolean,
        +lastMessaged: ?number
      |}>
    |}
  |}
|};
"
`;

exports[`generateFlowTypesFromJson Realistic examples api response should yield a sensible Flow type 1`] = `
"type UserAPIResponse = $ReadOnlyArray<{|
  +name: string,
  +age: ?number,
  +email: ?string
|}>;
"
`;

exports[`generateFlowTypesFromJson should find lists of objects, remove duplicates, unify all objects to a single object, and add maybe types to keys not present in all objects 1`] = `
"type TestType = {|
  +testList: $ReadOnlyArray<{|
    +test: string,
    +maybeProp: ?number,
    +anotherMaybeProp: ?string,
    +maybeObj: ?{|
      +test: string
    |},
    +nestedList: ?$ReadOnlyArray<{|
      +prop: string,
      +maybeProp: ?number
    |}>
  |}>
|};
"
`;

exports[`generateFlowTypesFromJson should handle nested arrays that are sometimes empty 1`] = `
"type TestType = {|
  +testList: $ReadOnlyArray<{|
    +test: string,
    +nestedList: $ReadOnlyArray<{|
      +prop: string,
      +maybeProp: ?number
    |}>,
    +maybeProp: ?number,
    +anotherMaybeProp: ?string,
    +maybeObj: ?{|
      +test: string
    |}
  |}>
|};
"
`;

exports[`generateFlowTypesFromJson should print JSON to a valid Flow type 1`] = `
"type TestType = {|
  +testString: string,
  +testBool: boolean,
  +testNum: number,
  +testNull: any, // FIXME: Type could not be determined
  +testObj: {|
    +str: string,
    +bool: boolean,
    +num: number,
    +not: any // FIXME: Type could not be determined
  |},
  +testObjList: $ReadOnlyArray<{|
    +str: string,
    +bool: boolean,
    +num: number,
    +not: ?any // FIXME: Type could not be determined
  |}>,
  +testPrimitiveList: $ReadOnlyArray<number | string>
|};
"
`;
